manual/src/tutorials/gadtexamples.etex

```diff
@@ -304,8 +304,149 @@ end
+  injectivity annotation, "type !'a succ", otherwise the definition of
+  "sized_list" is also rejected because "succ" is abstract.}
+\begin{caml_example}{verbatim}[error]
+open OtherTypeNat
+let of_singleton : ('a, zero succ) sized_list -> 'a = function
+  | Nil -> .
+  | Cons (v, Nil) -> v
+  | Cons (v, Cons _) -> .
+\end{caml_example}
+
+This problem also affects datatype definitions with the same
+constructor or field names, which could be alias of each other.
+\begin{caml_example*}{verbatim}[error]
+module DataTypeNat = struct
+  type zero = TypeNat
+  type 'a succ = TypeNat
+end
+open DataTypeNat
+
+let of_singleton : ('a, zero succ) sized_list -> 'a = function
+  | Nil -> .
+  | Cons (v, Nil) -> v
+  | Cons (v, Cons _) -> .
+\end{caml_example*}
+
+The recommended way to avoid this problem is to give a concrete
+definition to types such as "zero" and "succ", that are only meant to
+be used in GADT indices. (This is counter-intuitive, as we will never
+manipulate values of this type.) You could use:
+\begin{caml_example*}{verbatim}
+module TypeNat = struct
+  type zero = Zero
+  type 'a succ = Succ
+end
+\end{caml_example*}
+The types are now incompatibles as their constructor names are distinct.
+
+As an additional refinement, we can define those index types using
+using a "private" annotation, which clarifies that they are not meant
+to be used to build values:
+\begin{caml_example*}{verbatim}
+module TypeNat = struct
+  type zero = private Zero
+  type 'a succ = private Succ
+end
+\end{caml_example*}
+
+Remark: Usually "private" is meant to be used in interfaces, not in
+implementations; see the documentation of private types in
+\ref{s:private-types}. The constructors of a datatype that is
+\emph{defined} as "private" cannot be used to build values, so the
+type is effectively empty. This makes it more explicit that the types
+are only intended to be used as GADT indices.
+
+\subsection{s:abstract-types-current-module}{Legacy behavior: types in the current module}
+
+Early versions of OCaml, up to OCaml 5.3, had a special handling of
+types defined in the current module. (This special case does
+not apply to "succ" and "zero" above, as they are defined in
+a separate "TypeNat" submodule.) So the following definitions used to
+suffice to define "sized_list", as the compiler would accept that
+"zero" and "'a succ" are always distinct:
+
+\begin{caml_example*}{verbatim}
+type zero
+type 'a succ
+\end{caml_example*}
+
+The reasoning was that while in general abstract types could be
+defined as anything somewhere in the program, abstract types that we
+just introduced are not defined at all, so it is correct to assume
+that they are distinct. Similarly, datatypes defined in the current
+module were considered distinct even if they used the exact same
+constructors or field names.
+
+This special case made the system more complex and caused certain
+usability issues and type-checking issues. For example, GADT examples
```

Comment:
Which `type-checking issues`? And I would rather put the focus on the modularity issue to make it clearer which usability issue we are speaking off.